subsys/secure_storage/src/its/transform/aead.c

```diff
@@ -0,0 +1,128 @@
+				   const void *input, size_t output_size, void *output,
+				   size_t *output_len)
+{
+	psa_status_t ret;
+	psa_key_attributes_t key_attributes = PSA_KEY_ATTRIBUTES_INIT;
+	uint8_t key[CONFIG_SECURE_STORAGE_ITS_TRANSFORM_AEAD_KEY_SIZE];
+	psa_key_type_t key_type;
+	psa_algorithm_t alg;
+	psa_status_t (*aead_crypt)(const psa_key_attributes_t *attributes, const uint8_t *key,
+				   size_t key_size, psa_algorithm_t alg, const uint8_t *nonce,
+				   size_t nonce_length, const uint8_t *add_data,
+				   size_t add_data_len, const uint8_t *input, size_t input_len,
+				   uint8_t *output, size_t output_size, size_t *output_len);
+
+	secure_storage_its_transform_aead_get_scheme(&key_type, &alg);
+
+	psa_set_key_usage_flags(&key_attributes, operation);
+	psa_set_key_lifetime(&key_attributes, PSA_KEY_LIFETIME_VOLATILE);
+	psa_set_key_type(&key_attributes, key_type);
+	psa_set_key_algorithm(&key_attributes, alg);
+	psa_set_key_bits(&key_attributes, sizeof(key) * 8);
+
+	/* Avoid calling psa_aead_*crypt() because that would require importing keys into
+	 * PSA Crypto. This gets called from PSA Crypto for storing persistent keys so,
+	 * even if using PSA_KEY_LIFETIME_VOLATILE, it would corrupt the global key store
+	 * which holds all the active keys in the PSA Crypto core.
+	 */
+	aead_crypt = (operation == PSA_KEY_USAGE_ENCRYPT) ?
+		      psa_driver_wrapper_aead_encrypt : psa_driver_wrapper_aead_decrypt;
+
+	ret = secure_storage_its_transform_aead_get_key(uid, key);
+	if (ret != PSA_SUCCESS) {
+		return ret;
+	}
+
+	ret = aead_crypt(&key_attributes, key, sizeof(key), alg, nonce,
+			 CONFIG_SECURE_STORAGE_ITS_TRANSFORM_AEAD_NONCE_SIZE, add_data,
+			 add_data_len, input, input_len, output, output_size, output_len);
+
+	mbedtls_platform_zeroize(key, sizeof(key));
+	return ret;
+}
+
+enum { CIPHERTEXT_SIZE
+	= PSA_AEAD_ENCRYPT_OUTPUT_MAX_SIZE(CONFIG_SECURE_STORAGE_ITS_MAX_DATA_SIZE) };
+
+BUILD_ASSERT(CONFIG_SECURE_STORAGE_ITS_TRANSFORM_OUTPUT_OVERHEAD
+	     == CIPHERTEXT_SIZE - CONFIG_SECURE_STORAGE_ITS_MAX_DATA_SIZE
+		+ CONFIG_SECURE_STORAGE_ITS_TRANSFORM_AEAD_NONCE_SIZE);
+
+struct stored_entry {
+	psa_storage_create_flags_t create_flags;
+	uint8_t nonce[CONFIG_SECURE_STORAGE_ITS_TRANSFORM_AEAD_NONCE_SIZE];
+	uint8_t ciphertext[CIPHERTEXT_SIZE];
+} __packed;
+BUILD_ASSERT(sizeof(struct stored_entry) == SECURE_STORAGE_ITS_TRANSFORM_MAX_STORED_DATA_SIZE);
+
+/** @return The length of a `struct stored_entry` whose `ciphertext` is `len` bytes long. */
+#define STORED_ENTRY_LEN(len) (sizeof(struct stored_entry) - CIPHERTEXT_SIZE + len)
+
+struct additional_data {
+	secure_storage_its_uid_t uid;
+	psa_storage_create_flags_t create_flags;
+} __packed;
+
+psa_status_t secure_storage_its_transform_to_store(
+			secure_storage_its_uid_t uid, size_t data_len, const void *data,
+			psa_storage_create_flags_t create_flags, uint8_t stored_data
+			[static SECURE_STORAGE_ITS_TRANSFORM_MAX_STORED_DATA_SIZE],
+			size_t *stored_data_len)
```

Comment:
Actually the pattern is uniform, in line with the PSA Secure Storage API specification. It's as follows, in order:
- the current `size` (if the buffer is receiving data) or `length` (if the buffer already contains data) of the buffer
- the buffer
- if the buffer is receiving data, the output `length` i.e. how many bytes were placed in the buffer (so the argument is a pointer)

In this case there is no `size` because it's embedded as part of the buffer argument, between the square brackets. And `stored_data_len` is the output length.

Reply:
> in line with the PSA Secure Storage API specification

You are right. For example https://arm-software.github.io/psa-api/storage/1.0/api/api.html#c.psa_its_set sets `length` before `p_data`. Sorry for the comment then. I got confused by the fact that the convention for the crypto is different. In that case (ex: https://arm-software.github.io/psa-api/crypto/1.2/api/ops/sign.html#c.psa_sign_message) there's `input` and then `input_length`.

Anyway, this is not an issue then. Solved.

Reply:
Yeah, I saw that. A bit too bad that the two APIs are not aligned on this.